sys/linux/vnet.txt

```diff
@@ -738,24 +736,117 @@ ipv6_payload [
 # http://www.iana.org/assignments/tcp-parameters/tcp-parameters.xhtml
 
 include <net/tcp.h>
+include <net/mptcp.h>
+include <net/mptcp_v4.h>
+include <net/mptcp_v6.h>
 include <uapi/linux/tcp.h>
 
 tcp_option [
-	generic		tcp_generic_option
-	nop		tcp_nop_option
-	eol		tcp_eol_option
-	mss		tcp_mss_option
-	window		tcp_window_option
-	sack_perm	tcp_sack_perm_option
-	sack		tcp_sack_option
-	timestamp	tcp_timestamp_option
-	md5sig		tcp_md5sig_option
-	fastopen	tcp_fastopen_option
-	exp_fastopen	tcp_exp_fastopen_option
-	exp_smc		tcp_exp_smc_option
+    generic		tcp_generic_option
+    nop		tcp_nop_option
+    eol		tcp_eol_option
+    mss		tcp_mss_option
+    window		tcp_window_option
+    sack_perm	tcp_sack_perm_option
+    sack		tcp_sack_option
+    timestamp	tcp_timestamp_option
+    md5sig		tcp_md5sig_option
+    fastopen	tcp_fastopen_option
+    exp_fastopen	tcp_exp_fastopen_option
+    exp_smc	tcp_exp_smc_option
+    mptcp       tcp_mptcp_option
+] [varlen]
+
+tcp_mptcp_option [
+    generic     mptcp_generic_option
+    syn         mptcp_syn_option
+    synack      mptcp_synack_option
+    ack         mptcp_ack_option
+    capable     mptcp_capable_option
+    add_addr    mptcp_add_addr_option
+    mp_join     mptcp_mp_join_option
+    mp_flcose   mptcp_mp_fclose_option
+    remove_addr mptcp_remove_addr_option
 ] [varlen]
 
-tcp_option_types = TCPOPT_NOP, TCPOPT_EOL, TCPOPT_MSS, TCPOPT_WINDOW, TCPOPT_SACK_PERM, TCPOPT_SACK, TCPOPT_TIMESTAMP, TCPOPT_MD5SIG, TCPOPT_FASTOPEN, TCPOPT_EXP
+tcp_option_types = TCPOPT_NOP, TCPOPT_EOL, TCPOPT_MSS, TCPOPT_WINDOW, TCPOPT_SACK_PERM, TCPOPT_SACK, TCPOPT_TIMESTAMP, TCPOPT_MD5SIG, TCPOPT_FASTOPEN, TCPOPT_EXP, TCPOPT_MPTCP
+
+mptcp_sub_types = OPTION_TYPE_SYN, OPTION_TYPE_SYNACK, OPTION_TYPE_ACK, OPTION_MP_CAPABLE, OPTION_ADD_ADDR, OPTION_MP_JOIN, OPTION_MP_FCLOSE, OPTION_REMOVE_ADDR
+
+# https://tools.ietf.org/html/rfc6824#section-3
+mptcp_generic_option {
+     type    flags[mptcp_sub_types, int8]
+     length  int8
+     data    array[int8, 0:16]
+ } [packed]
+
+mptcp_capable_option {
+     type        const[TCPOPT_MPTCP, int8]
+     length      int8
+     version     int8:4
+     subtype     const[OPTION_MP_CAPABLE, int8:4]
+     flags       int8
+     sender      int64be
+     receiver    int64be
+ } [packed]
+ 
+ mptcp_add_addr_option {
+     type    const[TCPOPT_MPTCP, int8]
+     length  len[parent, int8]
+     version len[parent, int8:4]
+     subtype const[OPTION_ADD_ADDR, int8:4]
+     port    int16be
+     mac array[int8, 0:8]
+ } [packed]
+ 
+ mptcp_mp_join_option {
```

Comment:
Which figure in the RFC does this struct correspond to?

Reply:
i have to look into this mp_join again. according to the code, there is 

1. mp_join with syn
2. mp_join with synack and
3. mp_join with ack

Reply:
as there are 3 different mp_join constructions depending on the tcp flags SYN, SYNACK and ACK, could i just remove the struct for `mptcp_mp_join_option` because there are already 3 mp_join options (`mptc_syn_option`, `mptcp_synack_option`, and `mptcp_ack_option`)